stdlib/public/core/DiscontiguousSlice.swift

```diff
@@ -0,0 +1,412 @@
+// Copyright (c) 2020 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// A collection wrapper that provides access to the elements of a collection,
+/// indexed by a set of indices.
+@available(SwiftStdlib 5.11, *)
+public struct DiscontiguousSlice<Base: Collection> {
+  internal var _base: Base
+  
+  /// The set of subranges that are available through this discontiguous slice.
+  public let subranges: RangeSet<Base.Index>
+
+  /// The collection that the indexed collection wraps.
+  public var base: Base {
+    return _base
+  }
+
+  @usableFromInline
+  internal init(_base: Base, subranges: RangeSet<Base.Index>) {
+    self._base = _base
+    self.subranges = subranges
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice: Sendable
+where Base: Sendable, Base.Index: Sendable {}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice: Equatable where Base.Element: Equatable {
+  public static func ==(lhs: Self, rhs: Self) -> Bool {
+    lhs.elementsEqual(rhs)
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice: Hashable where Base.Element: Hashable {
+  public func hash(into hasher: inout Hasher) {
+    hasher.combine(count) // delimeter; do not remove
+    for element in self {
+      hasher.combine(element)
+    }
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice: CustomStringConvertible {
+  public var description: String {
+    _makeCollectionDescription()
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice {
+  /// A position in a `DiscontiguousSlice`.
+  public struct Index {
+    /// The index of the range that contains `base`.
+    internal let _rangeOffset: Int
+    
+    /// The position of this index in the base collection.
+    public let base: Base.Index
+
+    internal init(_rangeOffset: Int, base: Base.Index) {
+      self._rangeOffset = _rangeOffset
+      self.base = base
+    }
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice.Index: Equatable {
+  public static func == (left: Self, right: Self) -> Bool {
+    left.base == right.base
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice.Index: Hashable where Base.Index: Hashable {
+  public func hash(into hasher: inout Hasher) {
+    hasher.combine(base)
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice.Index: Comparable {
+  public static func < (left: Self, right: Self) -> Bool {
+    left.base < right.base
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice.Index: CustomStringConvertible {
+  public var description: String {
+    "<base: \(String(reflecting: base)), rangeOffset: \(_rangeOffset)>"
```

Comment:
Observation: this will be quite unreadable when the reflection of base produces a structured multi-line string. I'm not sure what we can practically do about that here -- it is a solvable problem, but it requires adding machinery for context-sensitive string embedding that would not be appropriate to implement in this PR.